### PR TITLE
Support octal, hex, and arbitrary radix numbers

### PR DESCRIPTION
**Pull request by jianlingzhong**
code is a bit ugly but should work:

```clojure
❯ build/jank repl
Bottom of clojure.core
clojure.core=> 071
57
clojure.core=> -071
-57
clojure.core=> 08
Read error (0 - 2): invalid number: char 8 are invalid for radix 8
clojure.core=> 08.9
8.9
clojure.core=> 08e2
800
clojure.core=> 0-7.1
0
clojure.core=> 0xabc
2748
clojure.core=> 0Xabc
2748
clojure.core=> -0xef
-239
clojure.core=> 0xg
Read error (0 - 3): invalid number: char g are invalid for radix 16
clojure.core=> 0x-a
Read error (0 - 4): invalid number: char - are invalid for radix 16
clojure.core=> 2r011110
30
clojure.core=> 8r71
57
clojure.core=> -8r71
-57
clojure.core=> 1r0
Read error (0 - 3): invalid number: radix 1 is out of range
clojure.core=> -36razxyu
-18473142
clojure.core=> -30razxyu
Read error (0 - 9): invalid number: char zxyu are invalid for radix 30
clojure.core=> 37rzb
Read error (0 - 5): invalid number: radix 37 is out of range
clojure.core=> 36r4.5
Read error (0 - 6): invalid number: char . are invalid for radix 36
clojure.core=> 36r-4.5
Read error (0 - 7): invalid number: char -. are invalid for radix 36
clojure.core=> 3e-4
0.0003
clojure.core=> 16r3e4
996
clojure.core=>  36r3e4
4396
clojure.core=> 3e2/3
Read error (0 - 3): invalid ratio
clojure.core=> 7r3e4
Read error (0 - 5): invalid number: char e are invalid for radix 7
clojure.core=>  0xe
14
clojure.core=> 16re3
227
clojure.core=> 16r3e
62
clojure.core=> 0x3e
62
clojure.core=>  0x3e3
995
clojure.core=> 16r3e3
995
```

### CODE REVIEW

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 1]
```diff
@@ -668,8 +868,8 @@ namespace jank::read
             pos++;
 
             /* Unescaped strings can be read right from memory, but escaped strings require
-             * some processing first, to turn the escape sequences into the necessary characters.
-             * We use distinct token types for these so we can optimize for the typical case. */
+                     * some processing first, to turn the escape sequences into the necessary characters.
+                     * We use distinct token types for these so we can optimize for the typical case. */
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:872]: This doesn't look right.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 2]
```diff
@@ -542,26 +681,79 @@ namespace jank::read
             }
 
             /* Tokens beginning with - are ambiguous; it's only a negative number if it has numbers
-             * to follow.
-             * TODO: handle numbers starting with `+` */
+                     * to follow.
+                     * TODO: handle numbers starting with `+` */
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:685]: Looks like an incorrect change.

Comment (jeaye) [reply compiler+runtime/src/cpp/jank/read/lex.cpp:685]: We can remove that TODO now, right? We handle numbers starting in `+`. Hmm, I don't see any tests for them though.

File: compiler+runtime/include/cpp/jank/read/lex.hpp [hunk 3]
```diff
@@ -151,14 +151,19 @@ namespace jank::read::lex
     processor(native_persistent_string_view const &f);
 
     result<token, error> next();
-    result<codepoint, error> peek() const;
+    result<codepoint, error> peek(native_integer const ahead = 1) const;
     option<error> check_whitespace(native_bool const found_space);
-    
+    native_bool is_valid_num_char(char const c) const;
+
```
Comment (jeaye) [inline compiler+runtime/include/cpp/jank/read/lex.hpp:157]: This doesn't need to be a member fn, from what I can see. It can just be a static fn in the cpp file.

File: compiler+runtime/include/cpp/jank/read/lex.hpp [hunk 4]
```diff
@@ -149,14 +149,19 @@ namespace jank::read::lex
...
     iterator begin();
     iterator end();
 
     size_t pos{};
-    /* Whether or not the previous token requires a space after it. */
+    native_integer radix{ 10 };
+    /* The 'r' used in arbitrary radix (prefixed with N and then r, where N is the radix (2 <= radix <= 36); */
+    /* e.g. 2r10101 for binary, 16rebed00d for hex) */
+    native_bool found_r{};
```
Comment (jeaye) [inline compiler+runtime/include/cpp/jank/read/lex.hpp:163]: Can these be stored as locals within the `next` fn? If so, let's prefer that.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 5]
```diff
@@ -524,15 +662,16 @@ namespace jank::read
                   return err(
                     error{ token_start, pos, "unexpected end of real, expecting exponent" });
                 }
-                break;
+                if(!is_letter(c) || (pos == token_start && file[token_start] != '0'))
+                {
+                  break;
+                }
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:668]: What're we checking here? Can you put whatever explanation you have into a comment, too, please? It's not obvious to me.

Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: added some comments for explanation.

File: compiler+runtime/include/cpp/jank/read/lex.hpp (hunk 4 above)
Comment (jianlingzhong) [reply compiler+runtime/include/cpp/jank/read/lex.hpp]: updated them as locals.

File: compiler+runtime/include/cpp/jank/read/lex.hpp (hunk 3 above)
Comment (jianlingzhong) [reply compiler+runtime/include/cpp/jank/read/lex.hpp]: udpated.

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 2 above)
Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: right, i didn't do anything regarding '+', so i'll keep it here.

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 1 above)
Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: udpated.

File: compiler+runtime/include/cpp/jank/read/lex.hpp [hunk 6]
```diff
@@ -150,14 +150,16 @@ namespace jank::read::lex
...
+    result<codepoint, error> peek(native_integer const ahead = 1) const;
     option<error> check_whitespace(native_bool const found_space);
 
     iterator begin();
     iterator end();
 
     size_t pos{};
-    /* Whether or not the previous token requires a space after it. */
+    /* The 'r' used in arbitrary radix (prefixed with N and then r, where N is the radix (2 <= radix <= 36); */
+    /* e.g. 2r10101 for binary, 16rebed00d for hex) */
```
Comment (jeaye) [inline compiler+runtime/include/cpp/jank/read/lex.hpp:161]: We've kept the comment from when we had a new member here.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 7]
```diff
@@ -440,81 +484,192 @@ namespace jank::read
...
+            auto r_pos{ pos }; /* records the 'r' position if one is found */
+            native_bool found_beginning_negative{};
+
+            if(auto const [c, l]{ peek().unwrap_or({ ' ', 1 }) };
+               file[token_start] == '-' && c == '0' && !found_slash_after_number)
+            {
+              contains_leading_digit = true;
+              if(auto const [f, l]{ peek(2).unwrap_or({ ' ', 1 }) }; f != ' ')
+              {
+                // auto const[n, _] {peek(3).unwrap_or({' ', 1})};
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:517]: Please remove this commented code, as well as the others in the various branches below.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 8]
```diff
@@ -440,81 +484,192 @@ namespace jank::read
...
+            {
+              contains_leading_digit = true;
+              // auto const[n, _] {peek(2).unwrap_or({' ', 1})};
+              if(auto const [f, l]{ peek().unwrap_or({ ' ', 1 }) }; f != ' ')
+              {
+                if(f != 'x' && f != 'X')
+                {
+                  radix = 8;
+                }
+                if(f == 'x' || f == 'X')
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:540]: Shouldn't this just be an `else`?

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 9]
```diff
@@ -542,27 +733,91 @@ namespace jank::read
...
+              if(found_r)
+              {
+                return err(error{ token_start,
+                                  pos,
+                                  "unexpected end of radix number, expecting more digits" });
+              }
+              return err(error{
+                token_start,
+                pos,
+                fmt::format("unexpected end of radix {} number, expecting more digits", radix) });
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:748]: Why do we have these two different returns? Does the last one not make sense if `found_r` is true?

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 10]
```diff
@@ -4,6 +4,8 @@
 #include <jank/read/lex.hpp>
 
 /* This must go last; doctest and glog both define CHECK and family. */
+#include <iostream>
+#include <clang/Basic/Diagnostic.h>
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:8]: Please put these at the top of the file. Also, the comment on line 6 is specifically for the doctest include.

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 11]
```diff
@@ -467,6 +487,104 @@ namespace jank::read::lex
...
+        CHECK(tokens
+              == make_results({
+                error{  0,  3,        "invalid number: char 3 are invalid for radix 2" },
+                error{  4,  8,       "invalid number: char z are invalid for radix 35" },
+                error{  9, 14,        "invalid number: char e are invalid for radix 8" },
+                error{ 15, 22,       "invalid number: char - are invalid for radix 19" },
+                error{ 23, 25, "unexpected end of radix number, expecting more digits" },
+                error{ 26, 29, "unexpected end of radix number, expecting more digits" }
+        }));
+      }
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:586]: I really like how you've captured both positive and negative tests here. Let's dig deeper into potential negative cases, since I think there are more. Off the top of my head, to get the ball rolling:

* `0r0` (is radix 0 supported?) -- In Clojure, no
* `1r1` (is radix 1 supported?) -- In Clojure, no
* `2048r0` (how high can radix actually go?) -- In Clojure, `36r0` is the highest
* `2.0r1` (what if radix is a real?)
* `2rr1` (what if `r` is specified multiple times?)
* `2r1r` (what if `r` is specified at the end of a valid number?)
* `2r.0` (what if a decimal is specified after `r`?)

Comment (jeaye) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:586]: Please also give the same treatment to octal and hex. I can help with some test cases if you need more suggestions.

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 8 above)
Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp:540]: i don't think so. if the string is "0)", then i don't want radix to become 16.

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 9 above)
Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp:748]: yeah, when found_r is true, we would not know what `radix` is yet until later so the msg could be confusing.

File: compiler+runtime/test/cpp/jank/read/lex.cpp (hunk 10 above)
Comment (jianlingzhong) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:8]: we actually don't need them. removed.

File: compiler+runtime/test/cpp/jank/read/lex.cpp (hunk 11 above)
Comment (jianlingzhong) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: updated with more unit tests.

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 12]
```diff
@@ -460,17 +478,332 @@ namespace jank::read::lex
...
+                error{ 21, 21, "unexpected character: ." },
+                error{ 22, 22, "expected whitespace before next token" },
+                token{ 22, 1, token_kind::integer, 2ll },
+        }));
+      }
+
+      SUBCASE("Arbitrary radix edge cases")
+      {
+        /* exceeds 64-bit integer max */
+        processor p{ "36r0123456789abcdefghijklmnopqrstuvwxyz" };
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:556]: Seems like this should be an error, until we support big numbers. The token we create is lossy.

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 13]
```diff
@@ -460,17 +478,332 @@ namespace jank::read::lex
...
+        native_vector<result<token, error>> const tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_tokens({
+                { 0, 39, token_kind::integer, 9223372036854775807ll }
+        }));
+
+        processor p2{ "2r1111111111111111111111111111111111111111111" };
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:563]: There's a neat edge case of a binary number with more bits than our integer. That should also be an error.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 14]
```diff
@@ -313,12 +325,52 @@ namespace jank::read
 
     static native_bool is_symbol_char(char32_t const c)
     {
-      return !std::iswspace(c) && !is_special_char(c)
+      return !std::iswspace(safe_cast_char32_t_to_int(c)) && !is_special_char(c)
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:328]: I don't think this is what we want. What we want is this:

* We have a `char32_t` and we want to check if it's a space (true or false)

With this change, we have this:

* We have a `char32_t` and we want to check if it's a space (true or false), but we throw an exception if it's larger than an int

Instead, let's just provide our own `iswspace` which takes a `char32_t`. The docs for it are here: https://en.cppreference.com/w/cpp/string/wide/iswspace Looks like a simple switch with 6 cases should work.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 15]
```diff
@@ -540,27 +737,95 @@ namespace jank::read
...
+                {
+                  invalid_digits.emplace_back(file[i]);
+                }
+              }
+              if(!invalid_digits.empty())
+              {
+                return err(
+                  error{ token_start,
+                         pos,
+                         fmt::format("invalid number: char {} are invalid for radix {}",
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp:803]: Let's clarify the error message here and fix the grammar by saying `invalid number: chars '{}' are invalid for radix {}`.

Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: updated

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 14 above)
Comment (jianlingzhong) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: updated to a static cast instead.

File: compiler+runtime/test/cpp/jank/read/lex.cpp (hunk 13 above)
Comment (jianlingzhong) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:563]: this one actually is not out of range.

File: compiler+runtime/test/cpp/jank/read/lex.cpp (hunk 12 above)
Comment (jianlingzhong) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:556]: updated to an error instead.

**Comment by jeaye** [comment]
Great work, Jianling!
